scripts/gen_csr_test.py

```diff
@@ -60,6 +60,35 @@ def write(self, new_field_val, csr_val):
             val_slice = new_field_val[self.pos[0]:self.pos[1] + 1]
             csr_val.overwrite(self.mask_bitarray & val_slice, self.pos[0])
 
+class LegalizeWriteCSRField:
+    def __init__(self, mask_bitarray, pos, read_only, legalize_py):
+        self.mask_bitarray = mask_bitarray
+        self.pos = pos
+        self.read_only = read_only
+        self.legalize_py = legalize_py
+        print(self.legalize_py)
+
+    def write(self, new_field_val, csr_val):
+        # only write if not read only
+        if not self.read_only:
+            val_slice = new_field_val[self.pos[0]:self.pos[1] + 1]
+
+            legal_new_val = self._legalize_val(val_slice.uint)
+            val_slice = bitarray(uint=legal_new_val,
+                length=self.pos[1] - self.pos[0] + 1)
+
+            csr_val.overwrite(self.mask_bitarray & val_slice, self.pos[0])
+
+    def _legalize_val(self, orig_val):
+        legalize_globals = {
+                'val_in' : orig_val,
+                'val_out' : orig_val
+        }
+
+        exec(self.legalize_py, legalize_globals)
+
+        return legalize_globals['val_out']
```

Comment:
Are `val_in/out` some kind of "global" variables to be used just for `legalize_py`? I am just asking out of interest, I have never seen this before.

Reply:
This is awesome @GregAC - I was looking into a fix for the very same issue.

Reply:
> Are `val_in/out` some kind of "global" variables to be used just for `legalize_py`?

Yes exactly. The `exec` call effectively runs the python in the `self.legalize_py` string in it's own interpreter environment. This sets up `val_in` and `val_out` as two global variables for it to use.